_posts/2025-05-05-date-histogram-optimizations.md

```diff
@@ -0,0 +1,216 @@
+
+This approach proved especially effective for minute-level aggregations, where traditional methods struggled to scale. As a result, daily and hourly aggregations saw up to 50x performance improvements, while minute-level aggregations improved by over 100x, with query times dropping from seconds to milliseconds.
+
+![Performance_Improvements_Final.png](../assets/media/blog-images/2025-05-05-date-histogram-optimizations/Performance_Improvements_Final.png)
+
+The following diagrams illustrate how documents are counted per histogram bucket using multi range traversal. Instead of restarting the traversal from the top for each bucket, it uses a two-pointer approach: one pointer tracks the current position in the tree, while the other tracks the active bucket. As the tree is traversed in sorted order, the algorithm checks whether the current value falls within the range of the active bucket. If it does, the document is collected; if the value exceeds the bucket’s upper bound, the pointer advances to the next bucket. This seamless transition between buckets avoids restarting the traversal and reduces redundant work. For example, if a node’s range (like 300–400) doesn't overlap with any target bucket, it’s entirely skipped. Similarly, nodes that are fully contained within a bucket (like 401–600) are directly counted without further exploration. This method is especially powerful when dealing with thousands of fine-grained buckets, such as minute-level aggregations, and dramatically reduces processing time by minimizing unnecessary operations.
+
+![MRT_Initial.png](../assets/media/blog-images/2025-05-05-date-histogram-optimizations/MRT_Initial.png)
+![MRT_Middle.png](../assets/media/blog-images/2025-05-05-date-histogram-optimizations/MRT_Middle.png)
+![MRT_Final.png](../assets/media/blog-images/2025-05-05-date-histogram-optimizations/MRT_Final.png)
```

Comment:
Seems there are no go-through explanation of these diagrams

Reply:
Yeah, @kolchfa-aws also added that feedback. Let me work on addressing that